= PuffinPlot User Manual
Pontus Lurcock <pont@talvi.net>
:numbered:
:xrefstyle: full
:title-logo-image: image::./figures/Fratercula_arctica.jpeg[width=300,align=center]
:doctype: book
:encoding: utf-8
:lang: en

// image:figures/Fratercula_arctica.jpeg[image,width=264]

== Introduction

This manual gives a practical guide to the usage of the PuffinPlot
palaeomagnetic data analysis program. Note that it does not elucidate the
principles of the analysis techniques themselves, and a working knowledge
of palaeomagnetic procedures is assumed. For excellent introductions to
palaeomagnetic analysis, see FIXME:TAUXE or FIXME:BUTLER.

If you use PuffinPlot in the creation of published work, please cite the
PuffinPlot paper (<<lurcock2012puffinplot,Lurcock and Wilson, 2012>>). See
<<sec:citing-puffinplot>> for further details.

PuffinPlot is free software, distributed under the GNU General Public
License. Please see the LICENCE file in the PuffinPlot archive for
details. PuffinPlot is copyright 2012–2020 by Pontus Lurcock. The
PuffinPlot website is located at http://talvi.net/puffinplot.

== Installation

This section describes the hardware and software requirements for
running PuffinPlot, and gives instructions for installing and running
it.

=== Requirements

PuffinPlot runs on the Java platform (version 8 or later), which is
preinstalled or freely available for a variety of operating systems,
including Windows, Linux, and Mac OS X. Java is generally easy to
install, but detailed instructions for installing it are beyond the
scope of this manual; please consult the documentation for your
operating system or visit http://www.java.com/ for details. (If you are
constrained to use an older version of Java on your system, you may
still be able to run an old version of PuffinPlot: version 1.03 of
PuffinPlot works on Java 7, version 1.02 on Java 6, and version 1.01 on
Java 5.)

=== Obtaining and installing PuffinPlot

You may download the latest version of PuffinPlot
from https://github.com/pont-us/PuffinPlot/releases.
PuffinPlot is distributed
as a single zip archive file
containing the program itself
(in two different variants, as detailed below)
and documentation.
The documentation consists of this manual
(in both HTML and PDF format)
and a JavaDoc folder,
containing documentation for PuffinPlot's source code.
The JavaDoc is only of interest
to users wishing to write scripts
which interface directly
with PuffinPlot's internal data structures (see <<sec:scripting>>).
<<tbl:archive-contents>> lists the contents of the PuffinPlot archive.
After downloading the archive,
move it to the folder where you would like PuffinPlot installed
and extract its contents there.

[[tbl:archive-contents]]
.The contents of the PuffinPlot archive file
[%autowidth.stretch,cols="<,<",options="header"]
|=======================================================================
|File |Description
|README |A text file briefly describing the archive.
|PuffinPlot.jar |The PuffinPlot program (suitable for all operating systems)
|PuffinPlot.app |The PuffinPlot program (tailored for Mac OS X)
|Manual |Folder containing this manual in PDF and HTML format
|JavaDoc |Folder containing internal documentation for PuffinPlot
|Examples |Folder containing example data files and scripts
|LICENCE |The licence under which PuffinPlot is distributed
|=======================================================================

==== All operating systems

The file PuffinPlot.jar in the archive is the PuffinPlot program itself,
as a cross-platform jar (Java archive) file. This file is recommended
for Linux, Windows, and other non-Mac operating systems; it will also
run on Mac OS X, but the specific Mac OS X version described below is
usually preferable. The details of starting PuffinPlot vary between
operating systems, but double-clicking on the file PuffinPlot.jar is the
most common method. In some cases it may be necessary to right-click on
the file and select _Open with Java runtime_ or some similar text from a
menu. PuffinPlot can also be started from a terminal prompt by setting
the current folder to the one containing PuffinPlot, and typing `java
-jar PuffinPlot.jar`.

*Note for Windows users.* Under Windows, other programs may occasionally
interfere with the Java platform; if you have the Java platform
installed on Windows but are still unable to run jar files such as
PuffinPlot.jar, the situation can usually be fixed by running the jarfix
program. At the time of writing, jarfix may be downloaded from
http://johann.loefflmann.net/en/software/jarfix/.

==== Mac OS X

The PuffinPlot archive also contains a special variant of the program
for Mac OS X. This variant is packaged as a standard OS X application
and conforms more nearly to OS X user interface standards than the
cross-platform version described in the previous section. It is named
'`PuffinPlot.app`', although the extension '`.app`' is usually hidden by Mac
OS X.

In the extracted archive contents, the PuffinPlot application is shown
as an icon representing a puffin (specifically, the Atlantic Puffin
_Fratercula arctica_). This application may be dragged into the
Applications folder, or any other convenient place.

PuffinPlot is started by double-clicking on its application icon. Recent
versions of Mac OS X include a security feature called '`Gatekeeper`',
which may prevent PuffinPlot from running. The Apple support note at
http://support.apple.com/en-gb/HT202491 gives instructions for
configuring Gatekeeper to allow the execution of an '`unidentified`'
application such as PuffinPlot.

== Basic usage

This section gives a brief introduction to the PuffinPlot data model and
to the features of its main window.

=== A note on keyboard shortcuts

PuffinPlot provides a large number of keyboard shortcuts for its various
capabilities; most of them involve holding down the Ctrl key while
pressing another key. On Mac OS X, these shortcuts use the command key
rather than the Ctrl key, in accordance with Apple user interface
guidelines. Throughout this manual, keyboard shortcuts will be given as
(for example) Ctrl-A. Users of Mac OS X should read these shortcuts as
command-A, etc.

=== Opening a file

To open a file, select Open… from the File menu. This will allow you to
choose one or more files to open from a standard file selection window.
You may also select an entire folder; in this case every file within the
folder will be opened and combined into a single suite. (On Mac OS X,
you can`'t choose a folder in the normal '`Open file`' dialog; you have to
select the Open folder… menu item instead.) You can also open a file,
folder, or group of files by dragging them onto the main PuffinPlot
window. *Note:* if you open multiple files or a whole folder, they must
have the same file type, the same treatment type (thermal,
alternating-field, etc.), and the same sample type (discrete or
continuous). Opening, for example, a mixture of Caltech and IAPD files,
or a mixture of discrete and continuous files, may result in errors
during file reading.

image:figures/annot-scrnshot.pdf[[[fig:screenshot]][fig:screenshot]The
main window of PuffinPlot with a discrete file loaded, showing the
default data plot layout. Data plots, information displays, and controls
are annotated. Note that several other data plots are available, but are
not shown by default, and that the layout may be freely rearranged by
the user. Data points above the 275C demagnetization step have been
hidden.,title="fig:"]
image:figures/annot-scrnshot.png[[[fig:screenshot]][fig:screenshot]The
main window of PuffinPlot with a discrete file loaded, showing the
default data plot layout. Data plots, information displays, and controls
are annotated. Note that several other data plots are available, but are
not shown by default, and that the layout may be freely rearranged by
the user. Data points above the 275C demagnetization step have been
hidden.,title="fig:"]

=== A tour of the main window

Figure 1 shows an annotated screenshot of PuffinPlot's main window with
a discrete sample data file open. The bulk of the window is devoted to
the data display itself, and shows the four most popular data display
methods for palaeomagnetic data: a table, a Zijderveld plot, an
equal-area projection, and a demagnetization-intensity biplot (which
also shows an overlaid magnetic susceptibility plot). The plots are
interactive, in that individual data points may be selected by clicking
or by dragging a rectangle with the mouse. The plot layout is
configurable: the sizes and positions of the plots can be changed, and
plots can be added to and removed from the display.

To the left of the main plot display area is the sample chooser, which
shows a complete list of the samples within the suite, with the
currently selected sample highlighted. Multiple samples may be selected
at once allowing functions such as PCA calculation to be performed _en
masse_. If a long core data file is loaded, the sample chooser is shown
as a vertical slider indicating depth rather than a list of sample
names.

Above the main plot area is the toolbar, which gives convenient access
to some of the most frequently used facilities of the program. From left
to right, these are: choosers for the currently displayed data suite,
the orientation correction, and the Zijderveld projection type; displays
of the sample and formation orientations and magnetic declination; and
buttons for three of the most commonly performed actions.

At the top of the window is the menu bar, providing access to all the
program's functions in a hierarchical manner. On Mac OS, the menu bar is
at the top of the screen rather than the top of the window, and includes
an extra menu at the left, entitled PuffinPlot.

=== Data model

image:figures/data-model.pdf[PuffinPlot's hierarchical data model. Each
layer (except the lowest) contains multiple instances of the following
layer.,title="fig:"] image:figures/data-model.png[PuffinPlot's
hierarchical data model. Each layer (except the lowest) contains
multiple instances of the following layer.,title="fig:"]

PuffinPlot uses a hierarchical data structure, with higher levels
containing multiple instances of each lower level. The structure is
summarized in Figure 2. At the top is the _suite_, which contains all
the data to be analysed as part of a particular study. For a discrete
specimen study, this will typically correspond to a section in the
field; for a long core study, it will correspond to a core. A suite is
initially created by opening one or more data files from a magnetometer;
it is saved as a file in PuffinPlot's own format. A suite can contain
multiple _sites_. A site corresponds to a set of samples taken from one
spot in a section (or from a particular range of depths). A site's
associated data can include such things as bedding attitude and
stratigraphic height, as well as calculated parameters such as the mean
palaeomagnetic direction for all the samples at the site. Sites are not
required: if no sites have been defined, samples are contained directly
within the suite.

Each site (or, if no sites are defined, the suite) contains multiple
_samples_. A sample corresponds to a small physical volume of rock. For
a discrete study, this will usually be a typical palaeomagnetic 25mm
cylinder or IODP cube sample. For long cores, it is the portion of the
core at a particular depth. The data associated with a sample consists
of information specific to this physical unit which does not change with
the application of demagnetization techniques — for example, a sample
code or name (or, for long cores, a depth), the field orientation of the
sample, and its volume. For discrete samples this data can also include
a tensor representing anisotropy of magnetic susceptibility, which is
imported separately from an Agico kappabridge datafile and collated with
the magnetization data by matching the sample names. The sample can also
contain calculated parameters, such as a direction fitted by principal
component analysis, or a best-fitting great circle.

Each sample contains multiple _demagnetization steps_. A _step_
represents a sample at a particular point during the treatment protocol.
Its associated data thus includes details of the treatment: the type
(thermal, AF, IRM, etc.) and parameters (temperature, field strength,
etc.). The data also includes the state of the sample itself — most
importantly, the measured magnetization vector. For thermal studies, the
magnetic susceptibility is usually also recorded after every heating
cycle, and is also stored as part of the step.

=== Main window features

This section describes the parts and functions of the main PuffinPlot
window, as shown in <<fig:screenshot>>.

==== Plot area

The plot area is the largest part of the window, and plots the data for
the current sample using various plots. By default, four plots are
shown: a demagnetization-intensity biplot, a Zijderveld plot, an
equal-area projection, and a table of demagnetization steps. The plots
can be moved and resized (see
<<sec:edit-layout>>). Other plots are also
available, and the preferences window can be used to control which plots
are displayed (see <<sec:preferences>>).

==== Sample chooser

The sample chooser sits at the leftmost edge of the main window, and
allows you to change the current sample (the one for which data is
plotted) and the set of selected samples (most of PuffinPlot's functions
operate on the currently selected samples). Often, the set of selected
samples will consist only of the current sample.

The sample chooser takes two forms, depending on whether the current
suite of data is for discrete samples or for a continuous long core
measurement.

*Using the discrete sample chooser*

The discrete sample chooser shows the names of the samples in the
current suite. The selected sample or samples are highlighted in a
different colour. The selected sample is the current sample, and its
data is displayed in the main plot area. If more than one sample is
selected, the first of the selected samples is the current sample.

To select a single sample, click on its name. To select a contiguous
range of samples, click at one end of the range, then hold down Shift
while clicking at the other end of the range. To select multiple,
non-contiguous samples, hold down Ctrl while clicking. To select all
samples, press Ctrl-A.

*Using the continuous sample chooser*

The continuous sample chooser is a vertical grey bar representing the
total length of the measured core, striped with horizontal white lines
representing the individual measurements at each depth. (If there are
too many measurements for all the requisite white lines to be displayed,
they are omitted.) A black triangle and line show the current depth;
this is the depth for which the data is displayed in the main window. If
there are selected samples, they are highlighted in red on the sample
chooser.

To select a single depth, click on the appropriate part of the sample
chooser. To scroll rapidly through a range of samples, click and drag
the mouse along the sample chooser. To select a range of samples, hold
down Shift, then click, drag, and release the mouse on the chooser.

*Keyboard shortcuts for sample selection*

Use Ctrl-B and Ctrl-N to change the current sample. Use Ctrl-A to select
all the samples in the current suite. You can also use the up and down
arrow keys to change the sample.

==== Toolbar [[sec:toolbar]]

The toolbar displays various data and provides several controls. From
left to right, these are:

Suite chooser::
  This shows the name of the current suite of data. If more than one
  suite of data has been opened, the suite chooser allows you to switch
  between them.
Orientation correction chooser::
  This chooser allows you to choose whether data is displayed in
  laboratory co-ordinates (uncorrected), in field co-ordinates,
  corrected for sample orientation (samp. corr.), or in tectonic
  co-ordinates, corrected for both sample orientation and bedding
  orientation (form. corr.).
Zijderveld vertical projection ('`V vs. …`')::
  This chooser controls the vertical projection used in the Zijderveld
  plot. The _y_ axis always corresponds to the vertical direction; the
  chooser controls the _x_ axis, which may correspond to North (V vs. N)
  or East (V vs. E). The third option, V vs. H, projects each data point
  separately, in the plane containing itself and the origin; this is
  sometimes referred to as a '`modified Zijderveld`' plot.
Zijderveld horizontal projection ('`N is up`'/'`W is up`')::
  This chooser controls the horizontal projection used in the Zijderveld
  plot. The upward direction on the plot can correspond to either north
  or west.
Sample orientation::
  (Samp). The first number is the azimuth of the sample orientation; the
  second is its either its dip angle or its hade, depending on the
  current setting in the user preferences (see
  <<sec:prefs-misc>>). By default, PuffinPlot
  uses dip angle rather than hade. For a long core, the azimuth and dip
  will usually be 0 and 90 respectively throughout the core.
Formation orientation::
  (Form). The first number is either the azimuth of the dip for the
  bedding, or its strike; the second is the dip angle. By default
  PuffinPlot uses the dip azimuth rather than the strike, but this can
  be changed in the preferences window (see
  <<sec:prefs-misc>>).
Magnetic declination::
  (Dev). This is the angle between magnetic north and true north at the
  sampling site. (It is abbreviated '`Dev`' (for '`deviation`') to avoid any
  possible confusion with the declinations of sample magnetizations.)
Select all::
  selects all the treatment steps in the current sample.
PCA::
  performs principal component analysis for the selected points of all
  the selected samples.
Clear::
  de-selects all the points in all the selected samples, and clears the
  results of any calculations done on them, such as PCA or great-circle
  analysis.

== Detailed usage

This section gives a methodical account of PuffinPlot's features.

=== Catalogue of functions

This section lists all the items in PuffinPlot's menus, giving a brief
description of the functionality associated with each one.

==== File menu [[sec:menu-file]]

This menu contains functions connected with opening, closing, and saving
files.

File → Open…::
reads one or more files of demagnetization data into PuffinPlot as a new
suite. See <<sec:file-types>> for details of
supported filetypes.

File → Open folder…::
reads a whole folder of data files into PuffinPlot. This menu item only
appears on OS X; on other operating systems, the Open… menu item also
allows selection of entire folders.

File → Open recent file::
is a submenu which contains the names of the last eight files which have
been opened in PuffinPlot, allowing them to be opened again with a
single click.

File → Save::
saves the current suite as a PuffinPlot file. If the suite was opened
from a PuffinPlot file or if it has been previously saved as a
PuffinPlot file, it will immediately be saved to that file. If no
PuffinPlot file is associated with this suite yet, a standard '`save
file`' dialog box will prompt you for a file name and location.

File → Save as…::
allows you to save the current suite to a different filename or
location.

File → Close::
closes the current suite, removing it from PuffinPlot's data display.

File → Export data::
is a submenu allowing the export of various kinds of data to CSV files.

File → Export data → Export sample calculations…::
saves a file containing all the data associated with individual samples.
<<tbl:export-sample>> describes the fields which make up the file.

[[tbl:export-sample,List of fields in exported sample data file]]
.List of fields in exported sample data file. Note that, in addition to the predefined fields, any custom user annotations (see <<sec:annotations>>) will also be exported in this file.
[cols="<,<",options="header",]
|=======================================================================
|Field name |Description
|Suite |Suite name
|Sample |Sample name (only present in discrete files)
|Depth |Depth in core (only present in long core files)
|NRM intensity (A/m) |NRM intensity in A/m
|MS jump temp. (degC) |For thermal demagnetization, the temperature step
at which the first jump in magnetic susceptibility occurs. A jump is
defined as a susceptibility of at least 2.5 times the previous value.

|Steps |Number of treatment steps for this sample
|PCA dec. (deg) |Declination of PCA direction (°)
|PCA inc. (deg) |Inclination of PCA direction (°)

|PCA MAD1 |The Maximum Angle of Deviation for the planar PCA fit; the
smaller the value, the more coplanar the points. See
<<sec:plot-types>> for more details.

|PCA MAD3 |The Maximum Angle of Deviation for the linear PCA fit; the
smaller the value, the more collinear the points. See
<<sec:plot-types>> for more details.

|PCA anchored |'`Y`' if the PCA fit was anchored; '`N`' if not

|PCA equation |The Cartesian equation of the PCA best-fit line

|PCA start (degC or mT) |Field (in mT) or temperature (in C) of first
demagnetization step used for PCA analysis

|PCA end (degC or mT) |Field (in mT) or temperature (in C) of last
demagnetization step used for PCA analysis

|PCA contiguous |'`Y`' if all steps between the first and last were
selected for PCA; '`N`' if any were omitted

|GC dec (deg) |the declination of the pole to the fitted great circle,
if any

|GC inc (deg) |the inclination of the pole to the fitted great circle,
if any

|GC strike (deg) |the strike of the plane of the fitted great circle, if
any

|GC dip (deg) |the dip of the plane of the fitted great circle, if any

|GC MAD1 |the MAD1 value for the great circle fit, indicating goodness
of fit (smaller is better). See
<<sec:plot-types>> for more details.

|GC npoints |the number of points used for the great-circle fit

|MDF half-intensity (A/m) |half of the NRM (in A/m)

|MDF demagnetization (degC or mT) |the treatment level at which the
intensity of the sample's remanence was reduced to half the NRM (in C or
mT). If half-intensity was not reached, this column contains 0.

|MDF midpoint reached |'`Y`' if magnetization intensity reached half the
NRM intensity during demagnetization; '`N`' otherwise

|Fisher dec. |Mean declination of treatment step directions (°)
|Fisher inc. |Mean inclination of treatment step directions (°)
|Fisher a95 |_α_~95~ of mean treatment step direction (°)
|Fisher k |_k_-value of mean treatment step direction
|Fisher nDirs |number of directions used to calculate mean
|Fisher R |length of sum of directions used to calculate mean
|AMS dec1 |declination of major axis of AMS tensor
|AMS inc1 |inclination of major axis of AMS tensor
|AMS dec2 |declination of intermediate axis of AMS tensor
|AMS inc2 |inclination of intermediate axis of AMS tensor
|AMS dec3 |declination of minor axis of AMS tensor
|AMS inc3 |inclination of minor axis of AMS tensor
|Initial MS |magnetic susceptibility at the first treatment step
|_[annotations]_ |Any user-defined annotations are also exported as part
of the sample export file. See
<<sec:annotations>> for details.
|=======================================================================

File → Export data → Export site calculations…::
saves a file containing all the data associated with sites.
<<tbl:export-site>> describes the fields which make up this file.

[[tbl:export-site]]
.List of fields in exported site data file
[cols="<,<",options="header",]
|=======================================================================
|Field name |Description
|Site |Name of site
|Samples |Number of samples at this site
|Fisher dec. |Mean declination of sample directions (°)
|Fisher inc. |Mean inclination of sample directions (°)
|Fisher a95 |_α_~95~ of mean sample direction (°)
|Fisher k |_k_-value of mean sample direction
|Fisher nDirs |number of directions used to calculate site mean
|Fisher R |length of sum of unit direction vectors
|GC valid |'`Y`' if the great-circle fit is valid, '`N`' otherwise. See
<<sec:prefs-misc>> for details on the validity
test and how it may be customized.

|GC dec. (deg) |Declination of great-circle direction (°)
|GC inc. (deg) |Inclination of great-circle direction (°)
|GC a95 (deg) |_α_~95~ for great-circle direction (°)
|GC k |_k_-value for great-circle direction
|GC N |Number of great circles used in great-circle fit
|GC M |Number of PCA directions used in great-circle fit
|GC R |length of sum of direction vectors in great-circle fit

|GC min points |the smallest number of treatment steps used to define
any of the great circles fitted at this site

|T1min |See note below
|T1max |See note below
|T2min |See note below
|T2max |See note below
|Lat (deg) |Site latitude
|Long (deg) |Site longitude
|VGP lat (deg) |VGP latitude
|VGP long (deg) |VGP longitude
|VGP dp (deg) |_dp_, the first semi-axis of the VGP confidence ellipse
|VGP dm (deg) |_dm_, the second semi-axis of the VGP confidence ellipse
|=======================================================================

Note on T1min, T1max, T2min, and T2max: these four parameters give the
ranges of demagnetization steps used to fit the circles. T1 denotes the
first (lowest) demagnetization step in a circle path for an individual
sample, and T2 the last (highest). T1min is the minimum of the T1 values
across all the circles for the site, and T1max the maximum. Similarly,
T2 denotes the last step used in a single circle, and T2min–T2max is the
range of its values across all the samples at a site.

File → Export data → Export suite calculations…::
saves a CSV file containing the Fisherian parameters for mean directions
calculated across the entire suite; see the documentation for the
Calculate → Suite means menu item
(<<sec:functions-calcs>>) for details.

File → Export data → Export multi-suite calculations…::
saves a CSV file containing the Fisherian parameters for mean directions
calculated across all the currently open suites; see the documentation
for the Calculate → Multi-suite means menu item
(<<sec:functions-calcs>>) for details.

File → Export data → Export IRM data…::
saves files containing IRM acquisition data. It produces a folder of
files, one for each sample in the suite. Each file is in tab-delimited
text format, and each line within the file contains the IRM field
strength and the magnetization intensity of the sample after application
of that field.

File → Export data → Create bundle…::
creates a _data bundle_ from the current suite. This is a zip archive
containing the suite data and analysis parameters, calculation results,
and other files useful for exploring or reproducing the results. Data
bundles are described in more detail in
<<sec:bundle>>.

File → Export graphics::
is a submenu with various options for exporting the plots for the
current and selected samples. See
<<sec:graphics-export>> for full details.

File → Import data::
is a submenu with functions to import additional data into the current
suite.

File → Import data → Append more palaeomagnetic data…::
reads a palaeomgnetic data file, like the Open… function in the main
File menu. The difference is that this function will append the data to
the current suite, rather than creating a new suite for it.

File → Import data → Import site locations…::
imports co-ordinates (latitude and longitude) for sites in the suite.
These co-ordinates are used to calculate virtual geomagnetic poles from
site directions. The location file should be in CSV (comma-separated
value) format with no header line. Column 1 contains the site name,
column 2 the latitude (degrees north from equator, negative for southern
hemisphere), and column 3 the longitude (degrees east of Greenwich).

File → Import data → Import AMS…::
imports AMS data from a file and adds it to the current suite. The file
must be in the .ASC format produced by the SAFYR and SUSAR programs
distributed with AGICO kappabridges. (AMS import has been tested with
data from SAFYR versions 1.5 and 2.6 and SUSAR version 4.0.) The AMS
data is assigned to the appropriate samples within the suite by matching
the sample names specified in the ASC file with the sample names for the
demagnetization data. If the AMS file contains data for samples not in
the suite, these samples will be created and added to the suite.
PuffinPlot reads the sample correction from the ASC file; it also reads
the F1 ('`foliation 1`') orientation and uses it as the formation
correction. If ASC data is being imported for existing samples, you can
choose either to keep the pre-existing sample and formation corrections,
or to replace one or both of them with the values read from the ASC
file. AMS data is not displayed by default; the equal-area plot of AMS
data can be activated from the Preferences window.

File → Page Setup…::
opens a window allowing you to change the paper size, orientation, and
margins for printing.

File → Print…::
opens a window allowing you to print the selected samples. Note that
_only_ the selected samples will be printed, so if you wish to print the
whole suite use Edit → Select all first. On most systems this will also
allow you to print to a PDF file; Windows users may need to install a
virtual PDF printer, such as CutePDF Writer or Bullzip PDF Printer, in
order to produce PDF files.

File → Print suite EA window…::
prints the contents of a separate window showing an equal-are plot of
sample/site directions through the suite; see
<<sec:menu-window>> for more details.

File → Print site EA window…::
prints the contents of a separate window showing an equal-area plot of
directions at the current site; see
<<sec:menu-window>> for more details.

File → Run JavaScript script…::
runs a user-specified external script written in the JavaScript
programming language. See <<sec:scripting>> for
details.

File → Run Python script…::
runs a user-specified external script written in the Python programming
language. See <<sec:scripting>> for details.

File → Preferences…::
opens the preferences window. See
<<sec:preferences>> for details. On Mac OS X,
this item is found on the PuffinPlot menu to the left of the File menu,
rather than on the File menu.

File → Quit::
terminates PuffinPlot. On Mac OS X, this item is found on the PuffinPlot
menu to the left of the File menu, rather than on the File menu.

==== Edit menu

The Edit menu contains functions related to the manipulation of data.

Edit → Edit layout::
allows you to reposition and resize the plots in the main display area.
See <<sec:edit-layout>> for details.

Edit → Reset layout::
resets the sizes and positions of all the plots to their default values.

Edit → Treatment steps::
is a submenu containing editing functions related to treatment steps
within a sample.

Edit → Treatment steps → Select all steps::
selects all the treatment steps in all the selected samples, excluding
treatment steps which have been hidden (see Hide steps below).

Edit → Treatment steps → Clear step selection::
de-selects all the treatment steps in all the selected samples.

Edit → Treatment steps → Copy step selection::
copies the selected treatment steps for the current sample to an
invisible '`clipboard`'. The selection can be pasted from the clipboard to
other samples (see Paste point selection), selecting the same
demagnetization steps in those samples. This is useful for selecting the
same treatment steps in a large number of samples without having to
manually select them for each sample.

Edit → Treatment steps → Paste step selection::
takes the selected treatment steps from the clipboard (see Copy step
selection) and selects the corresponding treatment steps in all the
selected samples.

Edit → Treatment steps → Hide steps::
hides all the selected treatment steps in all the selected samples. This
removes them from all the graphical plots (which will be rescaled to
avoid unnecessary blank space), but not from the data table in the main
plot window; on the data table, hidden treatment steps are marked with a
dash symbol (–) to their left. Hidden steps can be restored using the
Show all steps menu item.

Edit → Treatment steps → Show all steps::
restores all hidden treatment steps in all the selected samples.

Edit → Treatment steps → Merge duplicate steps::
A treatment step is considered a duplicate of another if they are both
in the same sample and have the same treatment type and level. This
function will replace each group of duplicate steps within the selected
samples with a single step. The magnetic moment of the replacement step
is calculated as the arithmetic mean of the moments of the duplicate
steps. Other treatment step data, such as magnetic susceptibility, is
taken from the first step in each group of duplicates.

Edit → Samples::
is a submenu containing editing functions related to whole samples.

Edit → Samples → Edit sample parameters…::
opens a window which allows you to change the sample volume, sample
orientation, formation orientation, and local magnetic declination for
all the selected samples. Each parameter type can be individually
changed without affecting the values of the others. For convenience,
sample orientation can be specified using either dip angle or hade; it
is possible but pointless to enter values for both of these parameters,
since one will overwrite the other. Similarly, formation orientation can
be specified using either dip azimuth or strike.

Edit → Samples → Set treatment type…::
opens a window which allows you to change the treatment type for the
selected samples. Normally, the treatment type will be automatically
determined on opening a data file. If the file does not specify the
treatment type, PuffinPlot will attempt to guess it. Set treatment type
allows you to set the treatment type explicitly in cases where
PuffinPlot guesses wrong, or where the wrong treatment type is given in
the data file.

Edit → Samples → Rotate/invert samples::
is a submenu allowing the demagnetization data from the selected samples
to be rotated 180 around a selected axis. Such functionality is rarely
required but can be useful, for example, when it is found that a sample
has been incorrectly oriented during measurement, or when converting
between different orientation conventions.

Edit → Samples → Rotate/invert samples → Flip samples around X axis::
rotates the magnetic moment values for all selected samples 180 around
the X axis.

Edit → Samples → Rotate/invert samples → Flip samples around Y axis::
rotates the magnetic moment values for all selected samples 180 around
the Y axis.

Edit → Samples → Rotate/invert samples → Flip samples around Z axis::
rotates the magnetic moment values for all selected samples 180 around
the Z axis.

Edit → Samples → Rotate/invert samples → Invert sample directions::
Inverts every magnetization vector in every selected sample. (Each of
the x, y, and z co-ordinates of each vector is negated; this corresponds
to a point reflection through the origin.)

Edit → Samples → Rotate/invert samples → Align core section declinations…::
operates on continuous suites made up of core sections with no absolute
azimuthal alignment. The alignment is carried out using PCA directions,
which must be calculated before carrying out the alignment. The top core
section is rotated so that the PCA direction at the top aligns with a
specified reference direction; for each core section below the top
section, the section is rotated so that the PCA direction at its top
aligns with the PCA direction at the bottom of the core section above
it. The number of samples averaged at the core ends to produce the '`top`'
and '`bottom`' directions can be specified.

Edit → Samples → Remove samples by depth…::
removes from the current suite all the samples outside a specified depth
range.

Edit → Samples → Remove samples by treatment type…::
removes from the current suite all the selected samples with a specified
treatment type. A sample will be removed if _any_ of its treatment steps
is of the specified type.

Edit → Samples → Merge duplicate samples…::
combines any selected samples that have the same name or depth. Each
merged sample will contain all the treatment steps in all the duplicate
samples; if any of the treatment steps are duplicated, they will be
merged as described for the Merge duplicate treatment steps function.

Edit → Samples → Rescale susceptibility…::
scales all the magnetic susceptibility values for the suite by a
specified factor. This is useful since magnetic susceptibility meters
typically do not report values in standard S.I. units. Note that, unlike
many PuffinPlot operations, the scaling is applied to the _entire
suite_, not just the selected samples.

Edit → Sites::
is a submenu allowing site names to be modified for samples in various
ways.

Edit → Sites → Set site name…::
allows a single site name to be set for all the selected samples.

Edit → Sites → Set sites from sample names…::
sets site names for the selected samples by taking specified characters
from the sample names. The characters to use are specified by a list of
comma-separated numbers and number ranges; for example, entering 1,3,5-8
would give each selected sample a site name composed of the first,
third, and fifth to eighth characters of the sample name, so that a
sample with the name FFQB0529.1 would get the site name FQ0529. The
table below gives some further examples.

[cols="<,<,<",options="header",]
|=====================================================
|Sample name |Character selection |Resulting site name
|FFQB0529.1 |1,3,5-8 |FQ0529
|FFQB0529.1 |4-6 |B05
|CiLpA-10-53 |1-2,5,7-8 |CiA10
|CiLpA-10-53 |3-4,9-11 |Lp-53
|=====================================================

Edit → Sites → Set sites by depth…::
sets site names in continuous suites using the depths of the individual
samples. PuffinPlot asks for a thickness value, and groups the samples
within the core into sites of that thickness, effectively '`slicing`' the
core into equally thick sites. Each site is named after the shallowest
depth within it.

Edit → Sites → Clear sites::
removes all site names in the current suite.

Edit → Suite::
is a submenu of functions which affect the whole suite.

Edit → Suite → Edit custom flags…::
allows you to add or remove user-defined flags for the suite; see
<<sec:annotations>> for details.

Edit → Suite → Edit custom notes…::
allows you to add or remove user-defined note categories for the suite;
see <<sec:annotations>> for details.

Edit → Suite → Discrete to continuous…::
lets you convert a discrete suite to a continuous one by replacing each
sample name with an associated depth. Selecting this item brings up a
dialog which lets you choose a CSV file to open. The CSV file must
consist of two columns. The first column contains sample names, and the
second column contains the associated depths. The CSV file must contain
a row for every sample in the suite.

==== Calculations menu [[sec:functions-calcs]]

The calculations menu provides facilities for calculating magnetic
parameters and directions. Note that most calculations operate on all
the selected samples, not just the current sample.

Calculations → Calculate PCA::
calculates a best-fitting line to all the selected points in all the
selected samples, using principal component analysis (FIXME:KIRSCHVINK).
The PCA direction
is projected onto the Zijderveld plot. If the '`PCA anchored`' menu item
(below) is ticked, the resulting PCA fit line will be anchored to the
origin; if '`PCA anchored`' is not ticked, the PCA fit will be unanchored.

Calculations → PCA anchored::
is a menu item which may be toggled on or off. When it is on, a tick
mark appears next to it, and PCA analyses are constrained to pass
through the origin. In general, it is appropriate to anchor the PCA if
the analysed points are known to represent the final demagnetization
component – that is, they are trending directly towards the origin, and
deviations from this path are known to be due to measurement noise
rather than an offset in the true magnetization vectors. Leaving PCA
unanchored allows analysis of a non-final component, provided that it is
sufficiently well separated from other components.

Calculations → Fisher by site::
calculates a Fisherian mean direction for each selected site using the
PCA directions of its samples.

Calculations → Fisher on sample::
calculates a direction for each selected sample using the Fisherian mean
of the directions for the selected treatment steps. PCA is almost always
preferable for establishing a reliable sample direction, but this
technique can sometimes be useful for determining a polarity from very
noisy data. Note that, at present, Fisherian sample directions are _not_
saved as part of the PuffinPlot file, although they can easily be
recalculated if the selected demagnetization points are not changed. The
Fisherian mean and associated parameters are, however, included in
exported sample data files (see
<<tbl:export-sample>>).

Calculations → Suite means::
calculates Fisherian means across all the selected samples. Two sets of
means are actually calculated: one set is calculated from the PCA
directions of individual samples, and the other from the site mean
directions (if any have been computed). Note that, if there are site
means computed by great-circle intersection, only those considered valid
are used; see <<sec:prefs-misc>> for details of
the validity test. If a site has both a valid great-circle direction and
a Fisherian direction calculated from PCAs, the great-circle direction
will be used. Each set of means consists of an overall mean and
individual means for the upper and lower hemisphere, to cater for data
sets containing reversals. Corresponding VGPs are also calculated. All
these means can be exported as a CSV file using the File → Export Data →
Export suite calculations… menu item. The overall means and VGPs (but
not the individual upper/lower hemisphere means) are also shown in the
Suite table plot if it is active. Note that (like most PuffinPlot
functions) this feature operates on all the _selected_ samples; to
calculate means for the entire suite you must first select all the
samples.

Calculations → MDF::
calculates the Median Destructive Field (or, for thermal
demagnetization, the Median Destructive Temperature) of the selected
samples. This is the field (or temperature) at which the intensity of
the sample's remanence has been reduced to half of its initial value.
Once calculated, it is displayed on the demagnetization-intensity plot,
and can be saved as part of the exported sample data. If the sample
never reaches half-intensity during treatment, the MDF is undefined; in
this case, it is not shown on the plot, and an MDF value of 0 is used in
the exported sample data file.

Calculations → Clear sample calculations::
de-selects all the points in all the selected samples, and clears the
results of any calculations done on them, such as PCA or great-circle
analysis.

Calculations → Clear sample PCAs::
clears any PCA calculations for the currently selected samples.

Calculations → Clear sample GC fits::
clears any great-circle fits on them for the currently selected samples.

Calculations → Fit great circle::
calculates and displays a best-fitting great circle for all the selected
points in all the selected samples.

Calculations → Great circles::
calculates a best-fitting direction for all the great circles fitted at
the sites corresponding to the currently selected samples, using the
algorithm of FIXME:MCFADDEN. For any sample which has a PCA direction but no fitted
great circle, the PCA direction will be used as an endpoint. The
great-circle fit is shown both in the Equal-area (site) plot in the main
window (if the plot has been activated in the preferences), and in a
separate window which is opened automatically.

Calculations → Clear site calculations::
clears the results of any calculations associated with the selected
_sites_ (as opposed to samples); at present, this amounts to clearing
the best-fit great-circle direction for each selected site.

Calculations → Multi-suite means::
calculates Fisherian means across all the samples in _all_ the currently
opened suites. the results are not plotted, but they are shown in a
pop-up window and can be saved using the File → Export data → Export
multi-suite calculations… menu item.

Calculations → AMS::
is a submenu containing various functions for AMS calculation.
PuffinPlot can show the results of statistical calculations on AMS
tensors, giving mean directions and confidence ellipses for the
principal axes by one of three methods; at present, however, these
calculations cannot be performed by PuffinPlot itself. Instead, it makes
use of two Python scripts from the PmagPy suite (FIXME:TAUXE), bootams.py and
s_hext.py. In order to calculate AMS statistics, these scripts must
first be installed on the computer running PuffinPlot, and the folder
containing them must be specified in the Preferences window. The PmagPy
programs may be obtained from http://earthref.org/PMagPy/. All AMS
calculations operate on the currently selected samples, except for Clear
AMS, which operates on the entire suite. FIXME and chapter 13 of FIXME give more
details of tensor statistics, particularly with regard to the
application of bootstrap methods.

Calculations → AMS → Calculate bootstrap AMS::
calculates bootstrap statistics using the bootams.py program, producing
Kent error ellipses which are shown on the AMS plot in the main window.

Calculations → AMS → Parametric bootstrap AMS::
calculates bootstrap statistics using the bootams.py program, producing
Kent error ellipses which are shown on the AMS plot. It differs from the
previous function in employing a parametric bootstrap, which can provide
more realistic confidence intervals for small numbers of samples on the
(often reasonable) assumption that measurement uncertainties are
normally distributed across the selected samples.

Calculations → AMS → Calculate Hext on AMS::
calculates FIXME:HEXT statistics using the s_hext.py program and displays the mean
directions and error ellipses on the AMS plot.

Calculations → AMS → Clear AMS calculations::
Clears any previously done bootstrap and Hext calculations.

Calculate RPI…::
calculates relative palaeointensity using two open suites. One suite
provides the NRM, the other provides the normalizing value from ARM or
magnetic susceptibility. RPI calculations are saved directly to a CSV
file; the RPI estimate can also be shown in the RPI plot.

==== Window menu [[sec:menu-window]]

This menu allows you to open or close auxiliary windows.

Window → Data table::
opens (or closes) a window showing all the demagnetization data for the
current sample as a table. This table is far more extensive than the
brief table displayed in the main window, and allows data to be selected
and copied to the clipboard so that it can be pasted into a spreadsheet
or text editor.

Window → Site equal-area plot::
opens (or closes) a window containing an equal-area plot for the current
site; the plot is created by selecting the Calculations → Fisher by site
or Calculations → Great circles menu items, and may be printed using the
File → Print site EA window… menu item. Note that the main display area
provides a similar plot; this window can be useful for a quick
inspection of site data at a larger scale without editing the main plot
layout.

Window → Suite equal-area plot::
opens (or closes) a window containing an equal-area plot of sample or
site directions across the whole suite; the plot is created by selecting
the Calculations → Suite means menu item, and may be printed using the
File → Print suite EA window… menu item. Note that the main display area
provides a similar plot; this window can be useful for a quick
inspection of suite data at a larger scale without editing the main plot
layout.

==== Help menu

Help → PuffinPlot website::
opens the PuffinPlot website using the default web browser.

Help → Cite PuffinPlot…::
opens a window containing information on the PuffinPlot paper (FIXME:CITE) and how to
cite it.

Help → About PuffinPlot::
displays some brief information about PuffinPlot, including the version.
On Mac OS X, this item is also present on the PuffinPlot menu.

=== Features

This section presents PuffinPlot's features in moderate detail.

==== Supported file types [[sec:file-types]]

PuffinPlot can read a number of commonly used palaeomagnetic data
formats:

PuffinPlot::
  (filename suffix `ppl`): PuffinPlot's own file format.
Old PuffinPlot format::
  (filename suffix `ppl`): an earlier version of the PuffinPlot format,
  produced by older releases of the program.
2G::
  (filename suffix `dat`): files produced by the '`Long Core`' program used
  with 2G Enterprises cryomagnetometers.
Caltech (CIT)::
  (filename suffix `sam`, with associated data files in same folder):
  format used at the Caltech Paleomagnetics Laboratory and supported by
  the '`Paleomag`' program of FIXME:JONES.
IAPD::
  (filename suffix `dat`): format used by the IAPD program and its
  successors, developed by T. H. Torsvik et al. and used at the
  Geological Survey of Norway. IAPD files contain an '`a95`' value for
  each treatment step which gives an indication of the measurement's
  reliability. PuffinPlot does not import these values, but will show a
  warning if any of them has a value of 5 or greater.
PMD (Enkin)::
  (filename suffix `pmd`): a text-based format used by the PMGSC program
  of R. Enkin et al., and supported by other paleomagnetic software
  including Paleomac and Remasoft. Not to be confused with the binary
  PMD format native to Paleomac.
UC Davis (old)::
  A format used for 2G cryomagnetometer measurements at UC Davis during
  the 1990s.
Zplot::
  (filename suffix `txt`): format used by the '`Zplot`' program developed by
  Steve Hurst at Woods Hole.
JR6::
  (filename suffix `jr6`): a format developed by AGICO and supported by
  the REMA6W and Remasoft programs, among others.
Sample directions::
  (filename suffix `txt`): a file of sample-level directional data. This
  lets you use PuffinPlot to analyse directions even when no
  demagnetization data is available (or when the data is not
  palaeomagnetic in origin). Input files must contain three fields per
  line, with no header lines: sample name or depth; declination in
  degrees; and inclination in degrees. Fields are separated by commas,
  tabs, or space characters.
Custom formats::
  This option imports data from a file format freely defined by the
  user, within certain limits. Custom formats allow PuffinPlot to read a
  large variety of textual, tabular data formats. See
  <<sec:import-data>> for details.

Note that there are different ways to read 2G data files: users are
encouraged to read <<sec:prefs-2g-import>>
and ensure that the preferences are correctly set for their needs. (In
particular, *magnetization vectors may be read incorrectly from 2G long
core files if the wrong import settings are used*, so it is important to
check the settings before importing data.)

In general, support for other file formats is straighforward to add, and
most tabular textual file formats can be opened using the '`File → Import
data…`' function.

==== Importing data in a custom format [[sec:import-data]]

When you open a file and select the Custom format filetype, PuffinPlot
opens a dialog box allowing you to describe a text-based, tabular file
format; once you have specified a format, you can choose one or more
files in that format, which PuffinPlot will then open. The file is
assumed to contain one magnetic measurement per line. The file format
specification consists of two parts: _General settings_ and _Column
definitions_.

*General settings*

This part describes parameters relating to the file as a whole, rather
than individual columns.

Number of header lines to skip::
  Data files may include extra data ('`header lines`') at the start of the
  file, most often a line containing textual descriptions of the
  columns. This field lets you specify how many lines PuffinPlot should
  ignore at the start of the file, letting it skip over them.
Measurement type::
  This specifies whether the files contain discrete or long core
  measurements.
Treatment type::
  This specifies the type of treatment applied to the samples – thermal,
  AF, IRM, etc.
Column separator::
  For files which do not use fixed-width columns (for example, CSV
  files), this drop-down lets you select the character used to separate
  the columns. If '`Use fixed-width columns`' is selected, the column
  separator is not used.
Use fixed-width columns::
  Tabular text files usually use one of two conventions for separating
  columns: either the columns have widths which differ from line to
  line, and are separated with a special character such as a comma or
  tab; or the columns have the same width in every line, and are padded
  out to this width with space characters when the contents are shorter
  than the column width. Select this tick box to specify that the file
  format has fixed-width columns. In this case, the column separator
  will be ignored and you must specify the widths of the columns (see
  below). If this box is not selected, the column widths are ignored.
Column widths::
  If your file format uses fixed-width columns, you must specify them
  here as a list of numbers separated by commas. For example, if you
  have six columns, with the first being ten characters wide and the
  rest eight characters wide, you would enter 10,8,8,8,8,8 in this box.
  If your file format does not use fixed-width columns, this box is
  ignored.

*Column definitions*

This part gives the column number for each item of measurement data to
be read.

Column no.::
  This is the number of the column to read; columns are numbered from
  left to right, starting at 1.
Column contents::
  This is the data to read from the specified column. Note that not all
  data types need to be specified for a file format; at minimum, it is
  sufficient to specify the demagnetization step and the three
  components of the magnetization vector (either as X, Y, and Z moments,
  or as declination, inclination, and magnetization). For any other data
  types, PuffinPlot will set reasonable default values if they are not
  specified in the file format.

==== Selecting points

For most of PuffinPlot's functions, the data points of interest must be
selected before anything can be done to them. You can select data points
simply by clicking on them; if you click on a selected point it will be
de-selected. Selected points are drawn in red to distinguish them from
the black unselected points. Note that the notional data point itself is
the thing being selected, not the visual representation that you click
on. Thus, if you click on a point in one plot, the corresponding point
in all the visible plots will also turn red, since they are visual
representations of the same treatmentStep.

Since data points are relatively small, clicking accurately on them can
be inconvenient. PuffinPlot offers two alternative selection methods to
alleviate this problem. Firstly, by holding down the Shift key, you can
select a point simply by left-clicking near it; holding Shift and
right-clicking will de-select nearby points instead. Secondly, you can
left-click, hold the button, and drag the pointer across the graph,
creating a rectangle. Every point within the rectangle will be selected.
Dragging a rectangle with the right button will instead de-select every
point within the rectangle.

==== Working with multiple samples

Since most PuffinPlot operations are automatically applied to all the
selected samples, repetitive analysis work can often be done
automatically using the Copy point selection feature. For example, if
you wish to apply a PCA to the 100–250C demagnetization range of a
series of 50 samples, it can be done in four quick steps:

1.  For the first sample, select the points corresponding to the
100–250C demagnetization range.
2.  Select the 50 samples using the sample chooser, keeping the first
sample as the current one.
3.  Use Copy point selection on the Edit menu (or press Ctrl-Shift-C to
select the corresponding points in all the selected samples.
4.  Select PCA from the Calculations menu, or press Ctrl-R.

This will immediately perform PCA on all 50 selected samples.

=== Plots and other data displays

==== Available plot types [[sec:plot-types]]

This section lists and briefly describes the available plot types in
PuffinPlot. Some of them are not displayed by default, but these may be
activated via the preferences window (see
<<sec:preferences>>). Note that the term '`plot`'
is used rather loosely in this manual to refer to any movable element
displaying data within the main window. Thus, the '`plots`' listed below
include textual elements such as legends and tables.

AMS::
is a lower-hemisphere equal-area plot of AMS data, if any has been
imported. Maximum, intermediate, and minimum anisotropy axes are shown
as squares, triangles, and circles respectively. If AMS statistics have
been calculated (see
<<sec:functions-calcs>>), the mean
directions and confidence ellipses are also shown.

Data table::
is a table in which each row represents one demagnetization step for the
current sample. The columns, from left to right, give the
demagnetization step, declination, inclination, intensity, and magnetic
susceptibility. Selected points are denoted by an asterisk (*) to their
left; hidden points are denoted by a dash (–) to their left.

Demag.::
is a plot of treatment level (in mT or C) versus intensity of
magnetization (in A/m), shown as a line of filled points. If magnetic
susceptibility measurements have been taken, they are overlaid on the
same plot as unfilled points, with the scale shown on the right of the
graph. If PuffinPlot cannot find any data describing the demagnetization
treatment (i.e. AF field strength or temperature), the X axis will be
labelled '`measurement number`', and the X values will correspond to the
sequence of the data in the file.

Equal-area (sample)::
is a Lambert azimuthal equal-area projection showing the directions of
the current sample's magnetization vectors. Successive points are
connected by great-circle segments. Points in the upper hemisphere are
shown as unfilled (white) and connected by solid lines; points in the
lower hemisphere are filled (black) and connected by dashed lines. If a
great circle fit has been calculated for the sample, it is shown on this
plot, and the pole to the great circle is shown as a triangle.

Equal-area (site)::
shows all the great circles fitted at the current sample's site, along
with a best-fit direction calculated by the method of FIXME . On this plot,
the calculated site direction is shown as a circle. Any sample PCA
directions are shown as diamonds. Demagnetization steps used for the
great-circle fits are shown as squares. Poles to the great circles are
shown as triangles.

Equal-area (suite)::
shows all the site means for the suite, and a Fisherian mean and 95%
confidence interval for them. If two polarities are present in the
suite, two means are calculated and shown. If no sites are defined,
individual sample PCA directions and their means are plotted instead.
Note that, for site means calculated by great-circle analysis, only
valid means are shown. See <<sec:prefs-misc>> for
details on how validity is determined.

NRM Histogram::
shows a histogram of NRM intensities across the whole suite.

RPI Plot::
plots the current RPI estimate (if any) against depth.

Sample parameters::
shows the results of PCA and/or great-circle fits for data in the
current sample. If neither of these calculations has been done, this
plot is invisible. When visible, it shows the inclination and
declination of the first principal component, which corresponds to a
least-squares linear fit. It also shows the maximum angular deviation
(MAD) values MAD1 and MAD3, which function as goodness-of-fit parameters
(smaller is better). The MAD1 value gives an indication of how nearly
the points lie in a single plane; the MAD3 value gives an indication of
how nearly they lie along a single line. PCA analysis and MAD values are
explained in FIXME and in section 9.7 of FIXME. The plot also show the Cartesian
equation of the PCA best-fit line. For the great-circle fit, the plot
gives the inclination and declination of one of the great circle's
poles, and the MAD1 value indicating the goodness of fit of the great
circle. (Note that this may be different from the MAD1 for the PCA fit,
since different sets of points may be used for the two fits.)

Sample parameter table::
shows a summary of parameters for each sample within the current site:
declination, inclination, and type of analysis ('`PCAa`' and '`PCAu`' for
anchored and unanchored PCA respectively, and '`GC`' for great circle).
For PCA analysis, the declination and inclination give the first
principal component; for great-circle analysis, they give the pole to
the circle. Clicking on a line within this table will show the
corresponding sample's data – in effect it works as an extra sample
chooser. If no sites are defined, this table shows the sample parameters
for the entire suite (or as many of them as will fit within the table's
current dimensions).

Site parameter table::
shows a summary of parameters for each site within the current suite.
The columns are:

Site::
  the name or identifier of the site
n::
  the number of samples at the site
PCA::
  the number of PCA analyses for samples at the site
GC::
  the number of great circles fitted for samples at the site
dec.::
  the declination of the site mean direction
inc.::
  the inclination of the site mean direction
a95::
  the _α_~95~ value of the site mean direction
R::
  the total length of the sum of the direction unit vectors
type::
  the type of analysis used: '`Fshr`' for FIXME:FISHER analysis on PCA directions; '`GC`'
  for great-circle analysis (FIXME:MCFADDEN). '`GC`' is suffixed with either '`v`' for valid
  or '`i`' for invalid. See <<sec:prefs-misc>> for
  details on the validity test and how it may be customized.

If both a valid great-circle direction and a Fisherian direction exist
for a site, only the great-circle direction will be shown. Clicking on a
line within the site parameter table will jump to the first sample of
the corresponding site.

Site parameters::
shows the site mean direction as calculated either by Fisher statistics
or by the great-circle technique of FIXME. It gives the mean inclination and
declination and the _α_~95~ and _k_
parameters. If no site mean direction has been calculated for the
current site, this plot is invisible.

Suite table::
is a table showing suite-level mean direction and VGP data. See the
entry on Site parameter table for explanations of the Fisher parameters
associated with each mean direction. The directions shown are: '`Site
dir`', the mean direction calculated from site mean directions; '`Sample
dir`', the mean direction calculated from sample mean directions; '`Site
VGP`', the mean of the VGPs of sites; and '`Sample VGP`', the mean of the
VGPs corresponding to individual sample directions. As with most
PuffinPlot operations, the parameters shown are calculated using only
the samples and sites selected at the time the calculation is done.

Ternary demag.::
is an experimental ternary plot designed to display data from triaxial
IRM demagnetization experiments conducted according to the method of FIXME.
The position of a point on the plot reflects the relative strengths of
the three axes of magnetization, which in turn correspond to high,
medium, and low coercivity components. The path produced by points at
successive demagnetization steps thus shows the relative effects of
thermal unblocking (and possibly thermal alteration) on these
components.

Title::
shows the name of the current sample and the current site, for a
discrete suite. For a continuous (long core) suite, it shows the current
depth. This plot also shows suite-level Fisher statistics calculated
over samples and over sites, if these are available (see the Calculate →
Suite means function in
<<sec:functions-calcs>>).

VGP map::
is a world map (using a Mollweide projection) showing the locations of
virtual geomagnetic poles (VGPs) for each site in the current suite.
VGPs can only be calculated for sites whose location has been set; See
_Import site locations…_ in <<sec:menu-file>> for
details on reading site locations from a CSV file.

VGP table::
is a table showing virtual geomagnetic pole (VGP) data for each site in
the current suite. The columns are: site name, _φ_ (site
latitude), _λ_ (site longitude), VGP _φ_
 (VGP latitude), VGP _λ_ (VGP
longitude), dp (first semi-axis of VGP confidence ellipse), and dm
(second semi-axis of VGP confidence ellipse). VGPs can only be
calculated for sites whose location has been set; See _Import site
locations…_ in <<sec:menu-file>> for details on
reading site locations from a CSV file.

Zplot::
is a Zijderveld plot, overlaying an orthographic projection in the
horizontal plane with an orthographic projection in a chosen vertical
plane. The vertical plane can be controlled using the chooser on the
toolbar; see <<sec:toolbar>> for details. The
horizontal projection is shown with filled points, and the vertical
projection with unfilled points. If a PCA fit has been calculated for
this sample, the two projections of the PCA lines are overlaid on the
plot in blue. (If the V vs. H vertical projection is selected, only the
horizontal projection of the PCA line is shown, since V vs. H
effectively uses a different vertical projection for each point.) The
appearance of the PCA lines can be changed using the Preferences dialog
(see <<sec:prefs-misc>>).

Zplot key::
is a legend for the Zijderveld plot, showing the interpretations of the
filled and unfilled points and giving the units in which the axes are
calibrated.

==== Arranging the plots [[sec:edit-layout]]

PuffinPlot allows the plots to be freely rearranged and resized within
the display area; they can also be switched on and off as required (see
<<sec:preferences>>). To arrange the plots,
select Edit layout from the Edit menu. This puts PuffinPlot temporarily
into a special mode where plots become moveable. A tick appears next to
the menu item, and the plots are overlaid by pale orange rectangles,
allowing you to manipulate them. Each plot is also annotated with its
name, which helps to identify plots that are not currently displaying
any data (e.g. the '`PCA directions`' display if no PCA has been
performed). To resize a plot, click and hold on an edge or corner of its
rectangle, then drag it to the desired size. To move a plot, click and
hold in its central area and drag it to the desired location. Plots may
be overlapped freely. When you click in an area where two or more plots
overlap, the smallest plot is treated as the '`topmost`', and this is the
one which will be moved or resized.

Once the plots are arranged to your satisfaction, click Edit layout on
the Edit menu again to untick the menu item and resume normal operation.

The plot layout is saved with your other preferences, and will be
retained if PuffinPlot is closed and restarted. You can restore the
original layout using the Reset layout menu item.

Sometimes it can be useful to save and restore different plot layouts.
This can be done using the Export preferences… and Import preferences…
menu items (see <<sec:menu-file>>).

=== The preferences window [[sec:preferences]]

The preferences window is divided into three tabs, with four action
buttons at the bottom.

==== The action buttons

Clear::
clears all changed preferences, resetting them to their default values.

Import::
sets your preferences from a file saved using Export preferences. See
the description of Export preferences for details.

Export::
saves your current preferences to a file. In conjunction with the Import
preferences feature, this allows you to transfer your preferences from
one computer to another. It also allows you to keep multiple sets of
preferences and switch between them as needed. Probably the most useful
application is to save different plot layouts for different sets of
data.

Close::
closes the Preferences window.

==== The 2G import tab [[sec:prefs-2g-import]]

This tab contains options connected with reading data from .DAT files
produced by the 2G '`Long Core`' software.

Read magnetization from.::
This setting controls whether PuffinPlot reads a sample's magnetic
moment from the fields giving the Cartesian components of the
magnetization vector (X/Y/Z) or whether it reads the polar
represantation (the declination, inclination, and intensity fields). The
Cartesian components are stored to higher precision in the file, so
using them is preferable when possible. However, if the Cartesian
components are used when reading a long-core file, they must usually be
corrected for the effective sensor lengths (see below). If the sensor
lengths are unknown, reading data from the polar fields can be a useful
fallback. When reading a file using the '`X/Y/Z`' setting, PuffinPlot
first looks for the '`X corr`', '`Y corr`', and '`Z corr`' fields to determine
the magnetization vector. If these are not present, it falls back to '`X
mean`'/'`Y mean`'/'`Z mean`', then to '`X intensity`'/'`Y intensity`'/'`Z
intensity`'.

SQUID sensor lengths.::
As described above, the Cartesian magnetization components in long core
files are not corrected for effective sensor length, which is determined
by the response function of each SQUID and must be determined
empirically when setting up the machine. To produce a magnetization
vector for long core files when using the '`X/Y/Z`' setting, PuffinPlot
corrects the SQUID readings for these configured sensor lengths when
opening the file.

Protocol::
gives the measurement protocol used in taking the readings. A protocol
is a particular sequence of empty tray measurements and sample
measurements in defined orientations, undertaken for each set of
measured samples. The tray and sample measurements are combined by
PuffinPlot as it reads the file, providing a more accurate, corrected
moment measurement for each sample. (Magnetic susceptibility
measurements, if present, are also automatically associated with the
preceding magnetic moment measurement or measurements.)
<<tbl:import-protocols>> describes the
available protocols.

*Warning: incorrectly set sensor lengths can produce erroneous data!*
The sensor lengths are not written to the 2G file, so they _must_ be
correctly set in PuffinPlot's preferences before importing 2G long core
data from the X/Y/Z vector components. (When importing discrete data, or
reading from the polar fields in the 2G file, sensor lengths are not
required.) If the sensor lengths are not known, use the
'`Dec/Inc/Intensity`' setting in the 2G import tab. Failing to correct for
sensor length is particularly dangerous because the deviations from the
correct vectors are often small enough not to be immediately obvious.
See Section 3 of FIXME for a fuller discussion of sensor response functions.

[[tbl:import-protocols]]
.Measurement protocols for 2G data files
[cols="<,<",options="header",]
|=======================================================================
|Protocol |Description
|NORMAL |No extra tray or sample measurements are made. Each measurement
run consists simply of measuring the samples once in their normal
orientation.

|TRAY_NORMAL |Before each sample-measurement run, an empty tray
measurement run is made. The input file thus consists alternately of
empty-tray lines and sample-measurement lines. Each tray measurement is
used to correct the subsequent sample measurement.

|NORMAL_TRAY |As TRAY_NORMAL, but the tray measurement is made after the
sample measurement rather than before it.

|TRAY_NORMAL_YFLIP |As TRAY_NORMAL, but adding an extra sample
measurement as a third step. In the extra measurement, the sample is
rotated 180 around its _y_ axis, so that the _x_ and _z_ measurements
are inverted. Combining these readings improves not only the precision
but also the accuracy of the magnetic moments measured on the _x_ and
_z_ axes, since any systematic bias should be cancelled out by the
inversion. For the _y_ axis, accuracy is not affected but precision is
improved by averaging the two measurements.

|TRAY_FIRST |This is the simplest tray correction: the tray is measured
once at the start, and all subseuent measurements are sample
measurements. PuffinPlot corrects each sample measurement using the
initial tray measurement.

|TRAY_NORMAL_IGNORE |This option reads a file measured using the
TRAY_NORMAL protocol, but (like TRAY_FIRST) makes all sample corrections
using the initial tray measurement, and ignores all subsequent tray
measurements. The main intended use for this option is to allow direct
comparison between the TRAY_FIRST and TRAY_NORMAL protocols, to avoid
the extra work of using the TRAY_NORMAL protocol on sample suites for
which it is unnecessary.
|=======================================================================

==== The Plots tab

This tab shows a list of plots which PuffinPlot can display. You can
control which plots are shown by ticking and unticking the boxes next to
the plot names. The plot types are detailed in
<<sec:plot-types>>.

==== The Misc. tab [[sec:prefs-misc]]

Label equal-area plots::
If this box is ticked, the equal-area plots (sample, site, suite, and
AMS) will be shown with text labels at the bottom right, making them
easier to distinguish from one another.

Label treatment steps::
If this box is ticked, each point on the demagnetization plots
(Zijderveld and equal-area) will be labelled with the appropriate
treatment step (AF intensity or temperature).

Label samples in site plots::
If this box is ticked, sample directions (PCA or Fisher) will be
labelled with the sample's name or depth in the site equal-area plot.

Label points in suite plots::
If this box is ticked, sample and site directions will be labelled with
the appropriate name or depth in the suite equal-area plot.

Highlight current sample/site::
If this box is ticked, the current sample and site data will be shown in
red on the site and suite equal-area plots and on the sample and site
parameter tables. This feature makes it easier to pick out the current
sample and site in the context of larger-scale analyses, and is useful
for exploring data, particularly in conjunction with the ability to jump
to a sample or site by clicking on its line in a parameter table.

Show site α95s on suite plot::
If this box is ticked, any site direction shown on the suite equal-area
plot will be plotted along with a projected small circle denoting the
_α_~95~ interval. The site _α_~95~
circles are plotted in blue, to distinguish them from the suite
_α_~95~s (plotted in black).

Bedding is vs. magnetic north.::
If this box is ticked, the bedding azimuth for formation orientation
correction is assumed to be relative to magnetic north, and a correction
is applied for local magnetic deviation. (The sample azimuth is always
assumed to be relative to magnetic north; if it is relative to
geographic north, a magnetic declination of zero can be specified.)

Demag. y-axis label::
allows to you customize the text which labels the _y_ axis of the
demagnetization-intensity plot.

PmagPy folder::
sets the location of the PmagPy programs. If you wish to do calculations
of AMS statistics within PuffinPlot, it is necessary to have the PmagPy
programs bootams.py and s_hext.py installed (see
<<sec:functions-calcs>> for details). This
box allows you to specify to PuffinPlot the folder in which the programs
are installed.

Font::
allows you to change the font used in the plots: enter a font family
name into the box. PuffinPlot must be restarted for the change to take
effect. If the specified font cannot be found, a default fallback font
is used.

Look and feel::
controls the appearance of PuffinPlot's windows and menus. (It has no
effect on the functionality of the program.) _Native_ gives an
appearance intended to harmonize with other programs on the operating
system on which PuffinPlot is running. _Metal_ and _Nimbus_ are
cross-platform appearances which will look the same on any operating
system. _Default_ will use the default appearance for Java programs on
the current operating system; in most cases this will be the same as
_Native_ for Mac OS X and Windows, and _Metal_ for Linux. Changes to
this option will not take effect until PuffinPlot is restarted.

GC validity::
allows you to customize the conditions under which a site mean
calculated by great-circle intersection is considered valid. The
validity test is used in several ways:

* When site data is exported to a CSV file, one of the exported columns
gives the result of the validity test.
* Only valid site mean directions are shown on the suite equal-area
plot.
* Only valid site mean directions are used when calculating the overall
mean direction for a suite.

The validity test takes the form of a logical expression in the syntax
of the JavaScript programming language, involving the following
variables:

M::
  the number of stable endpoints (PCA directions) used in the fit
N::
  the number of great circles used in the fit
a95::
  the _α_~95~ value (semiangle of the 95% confidence
  region)
k::
  the _k_-value (estimate for _κ_)

The most useful components for constructing validity expressions are the
comparison operators `<`, `<=`, `>=`, `>`, the logical operators `&&`
and `||` (corresponding to '`and`' and '`or`' respectively), and
parentheses. A typical expression might be

`a95<3 && k>5 && (M>=3 || N>=5)`

which means that a great-circle fit will be considered valid if it has
an _α_~95~ below 3, a _k_ above five, and
either at least three endpoints or at least five circles. The default
value of this setting is `true`, which causes all great-circle fits to
be considered valid. If there is an error in the expression (i.e. if it
isn`'t a valid JavaScript expression, or if it doesn`'t produce a
true/false value), it will be evaluated as false.

Zplot PCA display::
controls the manner in which PCA directions are shown on the Zijderveld
plot (see <<sec:plot-types>>). The available
settings are:

Full::
  PCA lines will extend right to the edges of the Zijderveld plot.
Long::
  PCA lines will be shortened by 10% from the '`Full`' length.
Short::
  PCA lines will only extend through the points used to calculate the
  PCA.
None::
  No PCA lines will be shown.

Sample orientation::
controls how sample orientation is displayed at the top of the main
window: it can be shown either as azimuth and dip angle, or as azimuth
and hade (the complement of the dip angle).

Formation orientation::
controls how formation orientation is displayed at the top of the main
window: it can be shown either as dip azimuth and dip angle, or as
strike and dip angle.

=== Annotations [[sec:annotations]]

Annotations are a feature allowing short, categorized notes to be added
to each sample in a suite. The categories can be freely chosen.
Annotations come in two varieties, _custom flags_ and _custom notes_.
Custom flags embody a true/false value and are intended to record
whether a sample fulfils some criterion – for example, '`messy`',
'`low-temperature alteration`', or '`multiple components`'. Custom notes are
intended for adding short items of information which are not
automatically inferred by PuffinPlot – for example, '`number of
components`' or '`behaviour type`'.

*Adding annotation categories*

Annotations categories defined by selecting the Edit custom flags… or
Edit custom notes… item from the Edit menu. This will show a window
allowing you to add, rearrange, or remove the annotation categories. If
an annotation category is removed, all annotations made within that
category will be lost.

*Using annotations*

When custom flags or notes have been defined, an extra panel appears at
the right-hand side of the main window. For each custom note category
there is a text box into which text may be typed. For each custom flag
category there is a tick box which may be selected or de-selected.

Annotations are saved with the rest of the data in the PuffinPlot file;
they are also exported in the sample data CSV file produced by the File
→ Export data → Export sample calculations menu item.

=== Exporting graphics [[sec:graphics-export]]

==== Introduction

PuffinPlot can export the displayed plots in several ways, for printing,
incorporation into documents, and editing by other programs. Two formats
are supported:

SVG (Scalable Vector Graphics)::
  is a widely supported format for the display and editing of vector
  graphics data. SVG files can be opened and edited by vector graphics
  programs such as Inkscape and Adobe Illustrator, and can be
  incorporated into documents by a variety of programs. The chief
  limitation of the SVG files exported by PuffinPlot is that they can
  only contain one page, corresponding to the currently displayed data.
PDF (Portable Document Format)::
  is a popular format for on-screen display and printing of all kinds of
  documents. PuffinPlot can produce multi-page PDF documents with one
  page for each selected sample. Many vector graphics programs can
  import PDF files, but since PDF is a format designed primarily for
  display rather than editing, the results may be inferior to those
  produced with SVG.

Both these formats are formally standardized; however, they are also
large and complex, and they are supported to varying extents by a huge
number of programs. For these reasons, compatibility problems can
sometimes occur. It is difficult to produce a file which will be
guaranteed to work well with any program on any operating system.
PuffinPlot attempts to mitigate this problem by offering a range of
different graphics export options, as detailed in the next section.

==== Graphics export options

All of these export functions may be found in the Export graphics
submenu of the File menu, except for the '`Print to PDF`' option.

Export SVG (Batik)::
  saves the current contents of the main data display as an SVG file
  using the Batik software library.
Export SVG (FreeHEP)::
  saves the current contents of the main data display as an SVG file
  using the FreeHEP software library.
Export PDF (iText)::
  produces PDF file using the iText software library. The resulting PDF
  file will use the current graph layout and will contain one page for
  each of the selected samples.
Export PDF (FreeHEP)::
  produces a PDF file using the FreeHEP software library. The resulting
  PDF file will use the current graph layout and will contain one page
  for each of the selected samples.
Print to PDF::
  is another way of producing a PDF file. Select Print… from the File
  menu, and select PDF as the destination printer. If the PDF option is
  not available, you will first have to install a PDF printer driver;
  please see your operating system documentation for details.

These options are to some extent redundant: SVG files produced using the
two menu items should appear practically identical, as should the three
varieties of PDF file. However, the internal structures of the files are
different, which is useful in improving compatibility with other
programs. If, for example, you find that another program has trouble
reading the SVG file produced using the Batik option, you may find that
if FreeHEP option produces better results.

=== Running PuffinPlot from the command line [[sec:command-line]]

In addition to its graphical user interface, PuffinPlot has a simple
command line interface, allowing some operations to be performed from a
text-based console or automated script. In summary, PuffinPlot's
available command-line arguments are:

....
usage: java -jar PuffinPlot.jar <options>
 -help                        print this message
 -installjython               download and install Jython
 -process <file>              process given ppl file and save results
 -script <file>               run specified script
 -scriptlanguage <language>   language for script (javascript or python)
 -withapp                     create a Puffin application (script mode
                              only)
....

-help::
  prints a summary of the available command-line arguments, as seen
  above
-installjython::
  downloads the Jython package and installs it locally, allowing
  PuffinPlot to run scripts written in Python
-process <file>::
  specifies a PuffinPlot (.ppl) file. The file is opened in PuffinPlot
  and all standard sample, site, and suite calculations are performed on
  the data. The results are automatically saved as CSV files into the
  folder containing the original PuffinPlot file.
-script <file>::
  runs the specified Python or JavaScript script. See
  <<sec:scripting>> for more details.
-scriptlanguage::
  can only be used in conjuction with `-script`, and specifies the
  language in which the script was written. Allowed values are `python`
  and `javascript`. The default value is `python`.
-withapp::
  can only be used in conjuction with `-script`. If this argument is
  provided, a PuffinPlot object will be created and assigned to a
  variable named puffin_app.

=== Scripting [[sec:scripting]]

PuffinPlot's graphical desktop interface is intended to be the primary
way to interact with the program. However, it is often useful to be able
to control a program using a scripting language, in order to extend its
capabilities, integrate it conveniently with other programs, or process
large amounts of data without manual intervention. The Java platform
upon which PuffinPlot is built supports a number of scripting languages
which can easily interface with PuffinPlot. Perhaps most usefully, an
implementation of the Python programming language – named Jython (FIXME:CITE) – has
been developed for the Java platform. Since Python is widely used in
scientific programming and scripting, and familiar to a large number of
scientists, this provides a convenient route for anyone wishing to
integrate PuffinPlot with other data processing steps. Using Jython,
PuffinPlot can be controlled either from a pre-written script, or
interactively from a command shell which accepts and executes commands
one at a time from the user. Jython scripts can also be run from within
PuffinPlot itself, making it easy to perform scripted operations on
currently open data suites. Jython is not distributed with PuffinPlot,
but PuffinPlot will automatically download and install it the first time
it is required. In addition to Python, PuffinPlot directly supports
scripting in JavaScript.

Scripting allows you to extend the functionality of PuffinPlot without
modifying the main program – for example, to perform extra processing on
your data. It also allows you to reuse parts of PuffinPlot as a
convenient library for other programs.

There are three basic ways to control PuffinPlot with scripting:

1.  Save a script as a file, and use Run Python script… or Run
JavaScript script… from the File menu to run it. This will give your
script access to any data which has already been loaded into PuffinPlot.
2.  Save a script as a file and run PuffinPlot from the command line,
specifying the name of the script as a parameter using the following
syntax: java -jar PuffinPlot.jar -script MyScriptName.py, where
MyScriptName.py is the filename of the script you wish to run. In this
case, the script will be run as soon as PuffinPlot starts. A JavaScript
script can be run in a similar way by specifying the language using
another command-line argument: java -jar PuffinPlot.jar -script
MyScriptName.js -scriptlanguage javascript,
3.  Use a scripting language interpreter separate from PuffinPlot. In
this case, you must download and install the language yourself. This
approach lets you use any language available for the Java platform, not
just Python or JavaScript. Additionally, many languages (including
Jython) provide an interactive console, allowing you to control
PuffinPlot by typing commands one at a time, rather than executing a
whole pre-written file.

For the first two techniques – when the script is run by PuffinPlot
itself – a variable called puffin_app is created, which represents the
currently running instance of PuffinPlot. (When running a script
directly from the command line, the -withapp argument must be specified
to create this variable.) This variable can be used, for example, to
gain access to any data already loaded into PuffinPlot.

<<fig:script-example-1>> shows a simple
script demonstrating the use of PuffinPlot from within an external
Jython interpreter (although it can also be run directly from within
PuffinPlot). The script opens a data file, calculates the mean NRM, and
produces a file containing a PCA direction for each sample. Note that
virtually all of PuffinPlot's data and functionality is available to the
Python script, so far more complex examples are possible.

*Internal documentation for PuffinPlot*

To write scripts interacting with PuffinPlot, some knowledge of its
internals is of course necessary. PuffinPlot comes with complete
documentation (JavaDoc) for all its accessible data structures. If more
detail is required, the source code is also freely available.

....
### Import the required libraries.
from net.talvi.puffinplot.data import Suite
from net.talvi.puffinplot.data import Correction
from java.io import File

### Create a Suite and read a data file into it.
input_file = File("example.ppl") # Specify an input file
suite = Suite("Example script")  # create a new Suite
suite.readFiles([input_file])    # Read the data into a Suite object
samples = suite.getSamples()     # Get a list of the Samples in the Suite

### Calculate and display the mean NRM.
total_nrm = sum([sample.getNrm() for sample in samples])
print total_nrm / suite.getNumSamples()

### Perform a PCA calculation for each sample.
for sample in samples:           # do this for each sample:
    sample.selectAll()           # select all points in the sample
    sample.useSelectionForPca()  # and mark them for use in PCA
suite.doSampleCalculations(Correction.NONE)    # perform PCA for each sample

### Save the results of the PCA calculation.
output_file = File("example-results.csv")
suite.saveCalcsSample(output_file)
....

....
# This script goes through all the data in the current suite.
# For any treatment step that doesn't have a magnetic susceptibility
# measurement, it sets the magnetic susceptibility to zero.

suite = puffin_app.getCurrentSuite()        # Find the current suite.

for sample in suite.getSamples():           # For every sample in the suite,
    for step in sample.getTreatmentSteps(): # and for every step in the sample,
        if not step.hasMagSus():            # if it doesn't have a m.s. value,
            step.setMagSus(0)               # set the mag. sus. to 0.
....

=== Creating a data bundle [[sec:bundle]]

PuffinPlot can export a suite as a data bundle, intended mainly as an
aid to reproducible research. It provides a straightforward way to
package up all the relevant data and analysis parameters. Using the
contents of a data bundle both the data and analysis techniques can be
browsed in PuffinPlot, and calculation results can be reused as they are
or regenerated from scratch. The data bundle is useful both for personal
storage of research results, and for public archiving or distribution,
for example as supporting material submitted with a manuscript for
publication.

The data bundle is a zip archive containing the following files:

PuffinPlot data file::
  This is simply a copy of the current .ppl data file, containing
  palaeomagnetic measurements and analysis parameters such as site
  definitions and demagnetization steps selected for PCA.
Sample, site, and suite calculations::
  exported as CSV files. If no sites are defined for the suite, the site
  calculations are omitted.
Calculation scripts::
  which can be used to run PuffinPlot to process the data file and
  reproduce the calculations automatically.
PuffinPlot jar file.::
  This is a software archive containing the PuffinPlot program itself.
  Including this file in the bundle means that only a Java environment
  is necessary to re-run or modify the calculations. However, its
  inclusion significantly increases the size of the bundle, so it may
  optionally be omitted.
README file::
  describing the contents of the archive and explaining how the data can
  be inspected and used.

== PuffinPlot file format

=== General information

PuffinPlot files use a text-based format in UTF-8 encoding. The line
separator is a single newline (character number 10). Fields within a
line are separated by tabs (character number 9). At the time of writing,
PuffinPlot uses version 3 of its file format. Versions 1 and 2 were only
used in pre-release versions of PuffinPlot, and may therefore be safely
ignored.

The main part of the PuffinPlot file is a sequence of lines containing
treatment step data, one step to a line. The sequence of data fields is
not fixed; rather, it is defined by a header line at the start of the
treatment step section.

=== Sections

1.  The file identifier line, consisting of the exact string
`PuffinPlot file. Version 3`
2.  The header line, containing a tab-separated sequence of field
identifiers. This line defines the interpretation of the treatment step
lines which follow.
3.  Any number of treatment step lines. These must contain the same
number of fields as the header line. The interpretation of each field is
determined by the field identifier string in the corresponding position
in the header line.
4.  A blank line, indicating the end of the treatment step section.
5.  A sequence of lines containing sample, site, and suite data. There
may be any number of these (including zero), and they may occur in any
order.

=== Treatment step fields

Treatment step fields are detailed in
<<tbl:treatment-step-fields>>.

[[tbl:treatment-step-fields]]
.Treatment step fields in the PuffinPlot file format
[cols="1,1,1,3",options="header",]
|=======================================================================
| Identifier | Type / Unit | Default | Description
|`DISCRETE_ID` | string | `UNSET` | The discrete sample identifier (sample name).

|`DEPTH` | unspecified | null |
The depth of this sample in a core. Note that PuffinPlot does not assume any explicit 
unit for depth: any saved or exported data will simply use the same depth values without any 
attempt to interpret them in a unit system.

|`RUN_NUMBER` | integer | −1 |
The sequential number of the magnetometer run which measured this treatment step.

|`TIMESTAMP` | string | `UNSET` |
The time at which this measurement was made. PuffinPlot doesn't process this field, 
so the format is undefined.

|`SLOT_NUMBER` | integer | −1 |
For discrete measurements on a multi-sample tray, the number of the tray slot containing 
the sample.

|`MEAS_TYPE` | string | `UNSET` | Measurement type. Must be `DISCRETE` or `CONTINUOUS`. This field is required.

|`X_MOMENT` | A/m | 0 |
The x component of the sample's magnetic dipole moment per unit volume in the
uncorrected sample co-ordinate system.

|`Y_MOMENT` | A/m | 0 |
The y component of the sample's magnetic dipole moment per unit volume in the
uncorrected sample co-ordinate system.

|`Z_MOMENT` | A/m | 0 |
The z component of the sample's magnetic dipole moment per unit volume in the
uncorrected sample co-ordinate system.

|`MAG_SUS` | unspecified | NaN |
The sample's magnetic susceptibility at this treatment step. PuffinPlot does not specify any unit for susceptibility and treats it as a relative value.

|`VOLUME` | cm^3^ | 10.8 | The volume of a discrete sample.

|`AREA` | cm^2^ | 4 | The cross-sectional area of a long core sample.

|`SAMPLE_AZ` | ° | NaN | The azimuth (i.e. down-dip direction) of a sample's field orientation.

|`SAMPLE_DIP` | ° | NaN | The dip of a sample's field orientation.

|`FORM_AZ` | ° | NaN |
The azimuth (i.e. down-dip direction) of the formation from which the
sample was taken.

|`FORM_DIP` | ° | NaN | The dip of the formation from which the sample was taken.

|`MAG_DEV` | ° | 0 | The magnetic declination at the sampling site, i.e. the declination of the magnetic field direction relative to geographic north.

|`TREATMENT` | string | `UNKNOWN` | The type of treatment applied to the sample. Must be one of `NONE`, `DEGAUSS_XYZ`, `DEGAUSS_Z`, `ARM`, `IRM`, or `THERMAL`.

|`AF_X` | tesla | NaN | The x component of an alternating magnetic field treatment.

|`AF_Y` | tesla | NaN | The y component of an alternating magnetic field treatment.

|`AF_Z` | tesla | NaN | The z component of an alternating magnetic field treatment.

|`TEMPERATURE` | °C | NaN | The peak temperature of a thermal treatment.

|`IRM_FIELD` | tesla | NaN | The intensity of an applied IRM treatment.

|`ARM_FIELD` | tesla | NaN | The intensity of the DC field in an ARM treatment.

|`ARM_AXIS` | string | `UNKNOWN` |
The axis along which an ARM was applied. Must be one of `AXIAL`, `NONE`, or `UNKNOWN`.

|`PP_SELECTED` | boolean | false | Whether this step is selected in PuffinPlot.

|`PP_ANCHOR_PCA` | boolean | false | Whether PCA calculations for this step should be anchored.

|`PP_HIDDEN` | boolean | false | Whether this step is hidden in the PuffinPlot GUI.

|`PP_ONCIRCLE` | boolean | false | Whether this step is included in great-circle fits.

|`PP_INPCA` | boolean | false | Whether this step is included in PCA fits.
|=======================================================================


=== Required fields

The only required treatment step field in a PuffinPlot file is
MEAS_TYPE. The minimal valid PuffinPlot file is thus:

[literal]
PuffinPlot file. Version 3⏎
MEAS_TYPE⏎


(Here, ⏎ represents the newline character.)
This file contains no lines beyond the header line, and therefore no
data. The minimal file containing data is:

[literal]
PuffinPlot file. Version 3⏎
MEAS_TYPE⏎
DISCRETE⏎

This file contains a single treatment step, which specifies no data
except that the sample is discrete. Loading this file into PuffinPlot
will create a suite with a single discrete treatment step with all other
fields set to their default values.

=== Suite lines

A suite line has SITE as the first field and follows one of the
following formats:

SUITE MEASUREMENT_TYPE <type> ::
  The measurement type of this suite. Must be one of DISCRETE or
  CONTINUOUS.
SUITE CUSTOM_FLAG_NAMES <flag-name> ...::
  A list of names for user-defined flags containing sample information.
  Each sample can store a true/false value for each flag.
SUITE CUSTOM_NOTE_NAMES <note-name> ...::
  A list of names for user-defined notes containing sample information.
  Each sample can store a string value for each note.
SUITE CREATION_DATE <date> ::
  The date and time at which the suite was created. The format is
  ISO-8601 with date, time, milliseconds, and timezone, e.g.
  2019-03-12T16:17:34.347+01:00.
SUITE MODIFICATION_DATE <date>::
  The date and time at which the suite was last modified. Format is as
  for CREATION_DATE.
SUITE ORIGINAL_FILE_TYPE <string> ::
  The type of the file(s) from which the data for this suite was
  originally read. Must be one of TWOGEE, ZPLOT, PUFFINPLOT_OLD,
  PUFFINPLOT_NEW, CALTECH, IAPD, UCDAVIS, DIRECTIONS, CUSTOM_TABULAR,
  PMD_ENKIN, JR6, or UNKNOWN.
SUITE ORIGINAL_CREATOR_PROGRAM <string> ::
  The name and version of the program which originally created the
  suite.
SUITE SAVED_BY_PROGRAM <string>::
  The name and version of the program which saved this file.

(For convenience, fields are shown separated with spaces in this
listing; in reality they are separated with tab characters.)

=== Site lines

A site line has SITE as the first field and a site identifier string as
the second field. The line follows one of the following formats:

SITE <site-id> HEIGHT <height> ::
  A decimal number representing the height of the site. Units are
  unspecified.
SITE <site-id> LOCATION <latitude> <longitude> ::
  Two decimal numbers giving the latitude and longitude of the site in
  degrees.

(For convenience, fields are shown separated with spaces in this
listing; in reality they are separated with tab characters.)

=== Sample lines

A sample line has SAMPLE as the first field and a sample identifier
(discrete sample name or depth in a continuous core). The line follows
one of the following formats:

SAMPLE <sample-id> CUSTOM_FLAGS <flag-value> ...::
  The values of any custom flags defined for this suite. The number of
  values must match the number of custom flags defined in the SUITE
  CUSTOM_FLAG_NAMES line. Possible values are true and false.
SAMPLE <sample-id> CUSTOM_NOTES <note-content> ...::
  The values of any custom note defined for this suite. The number of
  values must match the number of custom note defined in the SUITE
  CUSTOM_NOTE_NAMES line.
SAMPLE <sample-id> SITE <site-id> ::
  The identifier of the site which contains this sample.
SAMPLE <sample-id> IMPORTED_DIRECTION <declination> <inclination> ::
  The palaeomagnetic direction of this sample. In most cases this
  direction is not explicitly stored, but recalculated from the
  treatment step data when PuffinPlot loads the file. Using an
  IMPORTED_DIRECTION line allows PuffinPlot to deal with samples that
  have defined directions but lack treatment step data.

(For convenience, fields are shown separated with spaces in this
listing; in reality they are separated with tab characters.)

== Acknowledgements

=== Libraries

PuffinPlot makes use of several software libraries generously released
under liberal terms:

* Apache Batik for SVG export.
* Apache Commons for mathematical operations, command-line argument
processing, and compression.
* AppBundler to package PuffinPlot as a Mac OS X application.
* FreeHEP for graphics export.
* iText for PDF export.
* JAMA for matrix calculations.
* JUnit and Mockito for unit testing.
* Jython for scripting.
* OrangeExtensions for an improved user interface on Mac OS X.

=== Other software

PuffinPlot was created with the assistance of a cornucopia of free and
open source software; among the more significant tools are the Java
language and platform (along with many third-party Java libraries),
NetBeans, Ant, Ivy, Emacs, LaTeX (plus many LaTeX packages), TeX4ht, the
Gimp, and the Ubuntu operating system. I thank all who contributed to
these projects.

=== People

PuffinPlot was initially developed at the Otago Palaeomagnetic Research
Facility at the University of Otago, Dunedin, New Zealand. Several
colleagues provided valuable testing and feedback during PuffinPlot's
early development; in particular I thank Christian Ohneiser, Faye
Nelson, Claudio Tapia, and Bethany Fox for their time and effort.
Further improvements were made during revision of the PuffinPlot paper,
in response to thoughtful suggestions from Gary Acton and an anonymous
reviewer. Since its initial release, PuffinPlot has benefited from bug
reports and suggestions from Bill Phillips, Jonathan M. Glen, Fabio
Florindo, Eric Horsman, Terence Day, Yoichi Usui, Adrian Muxworthy, Russ
Burmester, Andrei Kosterov, Luca Lanci, Radchagrit Supakulopas, Cyril
Okpoli, Bugra Cavdar, Jay Shah, Kuo-Hang Chen, Ashleigh Murszewski,
Robert Coe, Ana Paula de Martini de Souza, Sébastien Wouters, and many
others.

=== Icon

PuffinPlot's icon, and the frontispiece for this manual, are from an
illustration by the Finnish artist Wilhelm von Wright (1810–1887),
published in _Svenska fåglar, efter naturen och på sten ritade_ (2nd
ed., 1929).

== Future development and bug reporting

PuffinPlot continues to be developed, albeit slowly and sporadically now
that its main functionality is complete and stable. Bugs in existing
functionality are fixed as quickly as possible when they are reported,
and new features are added as time permits. The PuffinPlot project,
including downloadable packages of all released versions and the
complete source code, is hosted at http://talvi.net/puffinplot. Bug
reports, feature suggestions, and feedback of all kinds are always very
welcome. Please send them to puffinplot@gmail.com.

=== How to report a bug

If you are reporting a bug, please be as specific and detailed as you
can. Try to include as many as possible of the following:

* The actions that triggered the bug.
* What you expected to happen.
* What actually happened (i.e. the unexpected or buggy behaviour).
* The PUFFIN-ERROR.txt error report file, if PuffinPlot crashed and
wrote one.
* A file containing your PuffinPlot preferences; you can produce this
file by opening the Preferences window from the File menu and clicking
the Export button at the bottom.
* A copy of the data file(s) you were working on when the bug appeared.
* If PuffinPlot produced incorrect output in a file, include a copy of
that file.
* If PuffinPlot shows incorrect output on the screen, include a
screenshot.
* The version of PuffinPlot that you`'re using.
* The operating system and Java version that PuffinPlot is running on.

The more detail you include, the easier it will be to find and fix the
bug quickly.

== Citing PuffinPlot [[sec:citing-puffinplot]]

PuffinPlot was introduced and described in a 2012 paper published in
_Geochemistry, Geophysics, Geosystems_. If you make use of PuffinPlot in
published work, please include the following citation:

_________________________________________________________________________________________________________________________________________________________________________________________________
Lurcock, P. C., and G. S. Wilson (2012), PuffinPlot: A versatile,
user-friendly program for paleomagnetic analysis, _Geochemistry,
Geophysics, Geosystems_, 13, Q06Z45, doi:10.1029/2012GC004098.
_________________________________________________________________________________________________________________________________________________________________________________________________

The PuffinPlot paper is open access, and can be read online or
downloaded by visiting http://dx.doi.org/10.1029/2012GC004098. A PDF of
the paper can also be downloaded directly from the Otago University
research archive at
https://ourarchive.otago.ac.nz/bitstream/handle/10523/3651/2012GC004098.pdf
.

For convenience, citation data is provided below in two formats commonly
used by bibliography management software.

=== BibTeX citation record

....
@article{lurcock2012puffinplot,
   author = {Lurcock, P. C. and Wilson, G. S.},
   title = {PuffinPlot: A versatile, user-friendly program for
     paleomagnetic analysis},
   journal = {Geochemistry, Geophysics, Geosystems},
   year = {2012},
   month = {Jun},
   day = {26},
   publisher = {AGU},
   volume = {13},
   pages = {Q06Z45},
   issn = {1525-2027},
   doi = {10.1029/2012GC004098},
   url = {http://dx.doi.org/10.1029/2012GC004098}
 }
....

=== RIS citation record

....
 TY  - JOUR
 T1  - PuffinPlot: A versatile, user-friendly program for paleomagnetic analysis
 A1  - Lurcock, P. C.
 A1  - Wilson, G. S.
 Y1  - 2012/06/26
 JO  - Geochemistry, Geophysics, Geosystems
 VL  - 13
 SP  - Q06Z45
 SN  - 1525-2027
 UR  - http://dx.doi.org/10.1029/2012GC004098
 DO  - 10.1029/2012GC004098
 PB  - AGU
....

== Release notes

=== Next release

* Demagnetization plot: include a tick and label at 0 on the x axis. If
there is no magnetic susceptibility data to plot, include a tick and
label at 0 on the y axis too.

=== PuffinPlot 1.4.1 release notes

This is a minor release which fixes some bugs in version 1.4 and adds
two new plot types.

==== Data manipulation

* Declination realignment can now align to a mean declination as well as
to a top declination.

==== Graphing and data display

* Added a VGP map plot.
* Added an RPI/depth plot.

==== Calculations

* PuffinPlot no longer calculates an MDF values for samples which never
reach half-intensity, since the MDF is undefined in this case.

==== Data import

* Added support for more PMD file format variants.

==== Bug fixes

* Fixed a bug which prevented the Mac OS X app from starting.
* Re-enabled section numbering in the manual (accidentally disabled in
version 1.4).
* Deleting the current sample previously caused an error; now fixed.

=== PuffinPlot 1.4 release notes

The release numbering scheme changed with the 1.4 release: 1.4 is the
immediate successor of 1.03.

==== Installation requirements

* PuffinPlot now requires Java 8 or higher.

==== Data manipulation

* '`Discrete to continuous`' feature implemented, allowing conversion of
discrete suites to continuous suites using a supplementary CSV file
containing a sample-to-depth map.
* Automatic realignment of declination data, allowing reconstruction of
a continuous declination record for continuous suites in which the core
sections are not azimuthally oriented with respected to each other.
* Allow user to remove samples outside a specified depth range.
* Allow user to remove samples with a specified treatment type.

==== Calculations

* RPI calculation implemented, using normalization to ARM, IRM, or
magnetic susceptibility.
* Arason-Levi inclination-only statistics implemented, but currently
only accessible via the API, not the GUI.

==== Scripting

* The Jython package is no longer bundled as part of PuffinPlot,
dramatically reducing the download size. PuffinPlot still retains Python
support: Jython is downloaded and installed automatically if and when it
is required.
* PuffinPlot now supports scripting in JavaScript as well as Python.
* The user-defined great circle validity check now uses Javascript
rather than Python. In practice, this only means a slight change in
syntax.
* Some classes and methods in the PuffinPlot API have been renamed to be
clearer and more descriptive. This has no effect on users who only use
the graphical interface, but scripts written for older PuffinPlot
versions may have to be updated to work with the new API.

==== User interface

* Files can now be opened by dragging them to the main window.
* Current suite name and '`unsaved`' indicator added to the title bar.
* When saving files, PuffinPlot now defaults to the last-used folder.
* Status bar added to main window. When the mouse pointer is over a data
point representing a demagnetization step, the status bar shows details
of this step.
* The '`Edit`' menu has been divided into more submenus to more
conveniently accommodate the newly added functions.

==== Graphing and data display

* Per-sample Fisher statistics are now shown in the sample parameters
legend.
* R parameter now shown to 4 decimal places in site and suite parameter
tables.

==== Data import

* PuffinPlot can now import the text-based PMD format developed by R.
Enkin and supported by several palaeomagnetic programs including PMGSC,
Paleomac, and Remasoft.
* PuffinPlot can now import the JR6 format developed by AGICO and
supported by programs including REMA6W, Remasoft, and Paleomac.
* When importing from a custom file format, PuffinPlot now sets default
values for the sample and formation corrections if these are not
specified in the file.
* When importing AMS data from an Agico ASC file, PuffinPlot can now
read a formation correction from the file. All possible orientation
conventions are now supported on ASC file import; if necessary,
orientations are converted to PuffinPlot's conventions using the
orientation parameters stored in the file. Optionally, existing sample
and formation corrections can be overwritten with values read from the
ASC file.
* Import from Caltech (CIT) files has been improved: demagnetization
levels for thermal treatment are now read correctly, as is the '`NRM`'
treatment code.

==== Data export

* Degree signs in exported data replaced with '`deg`', to accommodate
software that has trouble with non-ASCII character sets.
* Precision of exported parameters increased to four decimal places.
* PuffinPlot can now export a bundle containing both data and a
processing script, allowing analyses to be reproduced automatically.
* In exported site data, field strengths for minimum and maximum
great-circle steps are written in millitesla rather than tesla.

==== Miscellaneous bug fixes

* Custom flags and notes are now included in exported sample data.
* PDF export no longer produces an extra page.
* Display is now updated automatically when sites are edited.
* Formatting of PCA equations in exported CSV files is now
locale-independent, so will always use '`.`' rather than '`,`' as the
decimal separator.
* Custom file import used to ignore the sample volume field; it now
makes use of it if present, and defaults to a volume of 1 cm³ if it is
not present.
* In demagnetization / intensity plots with AF treatment type, data
points for magnetic susceptibility measurements (if present) could
sometimes be plotted with an incorrect x position. This has now been
corrected.

==== Developer notes

* The build process now uses the Ivy dependency manager to download
required libraries automatically.
* Several hundred unit tests have been added, mainly for the data and
calculation classes. This helps to verify the correctness of
PuffinPlot's data processing, and to avoid the introduction of bugs
during future development.
* PuffinPlot's version control has been migrated from Mercurcial to Git,
and the main repository from Bitbucket to GitHub. (The Bitbucket
repository has been retained as a mirror.)

=== PuffinPlot 1.03 release notes

==== Calculations

* Virtual geomagnetic pole calculation.
* Fisher-by-site calculations can be done on continuous data sets.
* Fisher analysis of demagnetization steps.
* PCA and GC fits can be cleared individually.
* R added to the available Fisher statistical parameters.

==== Data plotting

* Horizontal projection in Zplot supports west-upward orientation.
* Data points can be labelled with treatment step.
* Equal-area plots can be labelled to avoid confusion.
* Site equal-area plots now distinguish PCAs, demag steps, GC poles, and
site means.
* Current site and sample are highlighted in the relevant data tables
and plots.
* Sample directions can be annotated with their names in the site
equal-area plot.
* a95 added to site parameter table.
* More compact default plot layout.
* Treatment steps can be labelled with the treatment level.
* Added suite parameter table for mean directions and VGPs.
* Site alpha-95s can be shown in the suite equal-area plot.

==== Data import

* Selectable units for custom data import.
* More variants of the AGICO AMS file can now be imported.
* Direct import of sample directions.
* IAPD file import.
* Caltech file import.
* Better guessing of measurement type in 2G files.
* Site location data import (for use with VGP calculation).
* More palaeomagnetic data can now be appended to an existing suite.

==== Data export

* n and a95 are listed in sample parameter file.
* Great circle strikes and dips are included in exported data.

==== Bug fixes

* Exporting FreeHEP SVG graphics no longer disables anti-aliasing.
* PuffinPlot no longer crashes if an incorrect 2G protocol is specified.
* PuffinPlot will not save its own file over the original data file.
* Cleared PCA directions no longer reappear.
* Fixed potential crash during suite parameter export.
* Measurement types now checked for consistency when opening a file.

==== Documentation

* All new features are fully documented in the updated user manual.

==== Miscellaneous new features

* Site calculations automatically update when sample calculations
change.
* Site directions cleared automatically when all their sample directions
are cleared.
* Warning when closing a file or quitting with unsaved data.
* Clearer error messages and warnings when reading corrupted files.
* Treatment type can be set manually.
* Site definitions can be cleared.
* Treatment steps can be deselected by dragging a box.
* PuffinPlot will ask for confirmation before overwriting files.
* Native file open dialog is now used on Mac OS X.
* Added '`invert sample moment`' feature.
* '`Open file`' dialogs now remember the last used folder.
* Sample volume can be edited.

==== Other notes

* PuffinPlot now requires Java 7 or higher.
* PuffinPlot is now hosted on BitBucket, due to the imminent demise of
Google Code.

No release notes are available for PuffinPlot versions prior to 1.03.

[bibliography]
== References

 - [[[butler1992paleomagnetism]]] Butler, R. F. (1992). Paleomagnetism:
Magnetic Domains to Geologic Terranes. Blackwell Scientific, Oxford.

 - [[[fisher1953sphere]]] Fisher, R. (1953). Dispersion on a sphere.
Proceedings of the Royal Society of London. Series A, Mathematical and
Physical Sciences, 217:295–305.

 - [[[hext1963tensors]]] Hext, G. R. (1963). The estimation of
second-order tensors, with related tests and designs. Biometrika,
50(3-4):353–373. Available from:
http://biomet.oxfordjournals.org/content/50/3-4/353.abstract[http://biomet.oxfordjournals.org/content/50/3-_4/353.abstract].

 - [[[jones2002paleomag]]] Jones, C. H. (2002). User-driven integrated
software lives: “Paleomag” paleomagnetics analysis on the Macintosh.
Computers & Geosciences, 28:1145–1151.

 - [[[juneau2009jython]]] Juneau, J., Baker, J., Munoz, L. S.,
Wierzbicki, F., and Ng, V. (2009). The Definitive Guide to Jython.
Apress, New York.

 - [[[kirschvink1980least]]] Kirschvink, J. L. (1980). The least-squares
line and plane and the analysis of palaeomagnetic data. Geophysical
Journal of the Royal Astronomical Society, 62(3):699–718.

 - [[[lowrie1990identification]]] Lowrie, W. (1990). Identification of
ferromagnetic minerals in a rock by coercivity and unblocking
temperature properties. Geophysical Research Letters, 17(2):159–162.
Available from:
http://dx.doi.org/10.1029/GL017i002p00159[http://dx.doi.org/10.1029/GL017i002p00159].

 - [[[lurcock2012puffinplot]]] Lurcock, P. C. and Wilson, G. S. (2012).
Puffinplot: A versatile, user-friendly program for paleomagnetic
analysis. Geochemistry, Geophysics, Geosystems, 13:Q06Z45. Available
from:
http://dx.doi.org/10.1029/2012GC004098[http://dx.doi.org/10.1029/2012GC004098].

 - [[[mcfadden1988circles]]] McFadden, P. L. and McElhinny, M. W.
(1988). The combined analysis of remagnetization circles and direct
observations in palaeomagnetism. Earth and Planetary Science Letters,
87:161–172.

 - [[[roberts2006high]]] Roberts, A. P. (2006). High-resolution magnetic
analysis of sediment cores: Strengths, limitations and strategies for
maximizing the value of long-core magnetic data. Physics of the Earth
and Planetary Interiors, 156(3–4):162 – 178.

 - [[[tauxe2010paleomagnetism]]] Tauxe, L., Butler, R. F., Banerjee,
S. K., and Van der Voo, R. (2010). Essentials of paleomagnetism.
University of California Press, Berkeley.

 - [[[tauxe1998directions]]] Tauxe, L., Gee, J. S., and Staudigel, H.
(1998). Flow directions in dikes from anisotropy of magnetic
susceptibility data: the bootstrap way. Journal of Geophysical Research,
103(B8):17775–17790.
